lib/AST/DiagnosticEngine.cpp

```diff
@@ -644,18 +618,16 @@ static void formatDiagnosticArgument(StringRef Modifier,
           << FormatOpts.ClosingQuotationMark;
     else
       Out << '@' << Arg.getAsDeclAttribute()->getAttrName();
-    break;
-
-  case DiagnosticArgumentKind::VersionTuple:
+  } else if (Arg.isa<llvm::VersionTuple>()) {
     assert(Modifier.empty() &&
            "Improper modifier for VersionTuple argument");
     Out << Arg.getAsVersionTuple().getAsString();
-    break;
-  case DiagnosticArgumentKind::LayoutConstraint:
+  } else if (Arg.isa<LayoutConstraint>()) {
     assert(Modifier.empty() && "Improper modifier for LayoutConstraint argument");
     Out << FormatOpts.OpeningQuotationMark << Arg.getAsLayoutConstraint()
         << FormatOpts.ClosingQuotationMark;
-    break;
+  } else {
```

Comment:
Huh, I'm actually not happy about this being converted from a switch to a branch tree. We've now lost an important signal if we ever decide to add a new kind of formatted primitive type.

Reply:
I'm not all that happy with this either, but one of the benefits of using TaggedUnion is we can have DiagnosticArguments that don't have trivial copy constructors. For context, I'm trying to work towards a solution to https://bugs.swift.org/browse/SR-11166 and related type-qualification issues. My idea was to introduce a new type of diagnostic argument that includes a type and which components of it require qualification in a specific context, but I'm having trouble coming up with a design that doesn't require this change.

Reply:
Deciding whether a type should print qualified or not depends on the declaration context the error is emitted from. It's a layering violation for the diagnostic engine to make that call.

Reply:
What I had in mind was an argument type that looked something like this:
```
struct ContextualTypeDescription {
  Type Ty;
  std::string ContextuallyQualifiedDescription; // A printed representation of the type which is unambiguous in some context
  bool RequiresAnyQualification; // Whether any part of the type requires qualification in some context.
};
```

Then, there would be utilities to create a `ContextualTypeDescription` at the point of diagnostic emission, e.g. in CSDiagnostics, by looking through decl contexts and eagerly printing the type w/ necessary qualification. DiagnosticEngine would work exactly the same except it would  use the ContextuallyQualifiedDescription and RequiresAnyQualification flag to determine the final output instead of applying the current heuristics.

If that still seems like a bad idea, I can drop this PR.

Reply:
Can you package up the type and printing options instead?